### PR TITLE
[Relay] Add expr_visitor, fix expr_functor exponential blowup problem

### PR DESCRIPTION
**Pull request by MarisaKirisame**
Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/dmlc/tvm/blob/master/CONTRIBUTORS.md#reviewers).
@wweic @icemelon9 @vinx13 @junrushao1994 @zhiics can you guys review?

### CODE REVIEW

File: python/tvm/relay/expr_functor.py [hunk 1]
```diff
@@ -36,9 +36,8 @@ def __init__(self):
     # pylint: disable=no-else-return
     def visit(self, expr):
         """Apply the visitor to an expression."""
-        found = self.memo_map.get(expr)
-        if found:
-            return found
+        if expr in self.memo_map:
```
Comment (junrushao1994) [inline python/tvm/relay/expr_functor.py]: Try this
```python3
found = self.memo_map.get(expr, None)
if found is not None:
```

Comment (MarisaKirisame) [reply python/tvm/relay/expr_functor.py]: this exactly do not work.

Comment (MarisaKirisame) [reply python/tvm/relay/expr_functor.py]: the problem is get might return None

Comment (MarisaKirisame) [reply python/tvm/relay/expr_functor.py]: (because ExprVisitor store None in them)

Comment (junrushao1994) [reply python/tvm/relay/expr_functor.py]: I see. Thanks!

**Pull request merged**